lib/chromic_pdf/pdf/browser/channel.ex

```diff
@@ -50,34 +47,117 @@ defmodule ChromicPDF.Browser.Channel do
+      Logger.error("""
+      ChromicPDF received an 'Inspector.targetCrashed' message.
+
+      This means an active Chrome tab has died and your current operation is going to time out.
+
+      Known causes:
+
+      1) External URLs in <link> tags in the header/footer templates cause Chrome to crash.
+      2) Shared memory exhaustion can cause Chrome to crash. Depending on your environment, the
+         available shared memory at /dev/shm may be too small for your use-case. This may
+         especially affect you if you run ChromicPDF in a container, as, for instance, the
+         Docker runtime provides only 64 MB to containers by default.
+
+         Pass --disable-dev-shm-usage as a Chrome flag to use /tmp for this purpose instead
+         (via the chrome_args option), or increase the amount of shared memory available to
+         the container (see --shm-size for Docker).
+      """)
+    end
+  end
+
+  # -------- Task execution ----------
+
+  # "Runs" the protocol processing until done or await instruction reached.
+  defp handle_run_protocol(task, %{conn: conn} = state) do
+    case Protocol.run(task.protocol, &Connection.dispatch_call(conn, &1)) do
+      {:halt, result} ->
+        GenServer.reply(task.from, result)
+        state
+
+      {:await, protocol} ->
+        push_to_waitlist(state, %{task | protocol: protocol})
+    end
+  end
+
+  # Removes matching task from waitlist and returns protocol.
+  defp handle_cancel_protocol(ref, state) do
+    pop_from_waitlist(state, fn task ->
+      if task.ref == ref do
+        task.protocol
+      end
+    end)
+  end
+
+  # Removes task for which protocol can consume incoming chrome message from waitlist and
+  # passes it back to `handle_run_protocol` for further processing and re-enqueueing.
+  defp handle_chrome_message(msg, state) do
+    case pop_from_waitlist(state, &match_chrome_message(&1, msg)) do
+      {nil, state} ->
+        state
+
+      {task, state} ->
+        handle_run_protocol(task, state)
+    end
+  end
+
+  # Matches message against single protocol and updates task on match.
+  defp match_chrome_message(task, msg) do
+    case Protocol.match_chrome_message(task.protocol, msg) do
+      :no_match -> false
+      protocol -> %{task | protocol: protocol}
+    end
+  end
 
-    {:noreply, update_protocols(state, protocols)}
+  # -------- State management --------
+
+  defp push_to_waitlist(%{waitlist: waitlist} = state, task) do
+    %{state | waitlist: [task | waitlist]}
+  end
+
+  # Pops a value for which predicate is truthy from waitlist, returns predicate result.
+  defp pop_from_waitlist(%{waitlist: waitlist} = state, fun) do
+    {matched, waitlist} = do_pop_from_waitlist(waitlist, fun, [])
+    {matched, %{state | waitlist: waitlist}}
   end
 
-  defp update_protocols(state, protocols) do
-    %{state | protocols: Enum.reject(protocols, &Protocol.finished?(&1))}
+  defp do_pop_from_waitlist([], _fun, acc), do: {nil, acc}
```

Comment:
```suggestion
  defp pop_if([], _fun, acc), do: {nil, acc}
```
it actually has nothing to do with a waitlist (would ne a great addition to the `List` module :D

Reply:
Ahh realized that it is in fact not `pop_if` but returns the predicate.... 🙈 forget it